vm/vm.go

```diff
@@ -280,6 +280,20 @@ func (vm *Engine) validateStateTransition(virtualTx *wire.MsgTx) error {
 	}
 
 	for i, witness := range vm.newAsset.PrevWitnesses {
+		// TODO(george): Dummy assignment, need to figure out how to get
+		// the sighash byte from the signature.
+		sighashType := txscript.SigHashType(witness.TxWitness[0][33])
```

Comment:
Convention is: If the signature is exactly 64 bytes, it's an implied `SigHashDefault`. If it's 65 bytes, it is a different sighash and the 65th byte is the sighash byte.
We probably also need some logic to find out which entry of the witness is an actual signature.

Hmm... and that just get me thinking. Maybe we need a different way of transporting the sighash flag than by appending it to the witness. Because we look at the sighash on the level "vPSBT -> dummy TX". But the underlying Taproot execution VM will look at the sighash on the level "dummy TX -> txsighash", where we'll probably want to always use SigHashDefault. So the actual signature shouldn't contain the sighash as on the lowest level we always want it to be SigHashDefault. But when validating the asset, we need to know what the "vPSBT -> dummy TX" sighash was. So we need to transport it somewhere.

Or maybe this is an argument for actually mapping the vPSBT inputs/outputs directly into dummy TX inputs/outputs, then we can use the same sighash logic on both levels.

Reply:
Right, thanks for writing this up, seems like there's loss of information by using the 1-1 dummy tx if we want to use more sighash flags.

One thing that I hadn't realized is that when the vPSBT is no longer around, a future verifier will have to rely on the [`ChallengeWitness`](https://github.com/lightninglabs/taproot-assets/blob/main/proof/proof.go#L297) field of `Proof` in order to check that a past transfer was valid. That `ChallengeWitness` currently is the single 1-1 dummy tx (which relied on tap always using "sighash all"). This means that the ephemeral dummy txs constructed by each vInput in order for the backend to sign will no longer be reconstructible by the verifier in order to check the signature.

As you mention, we can expand the `ChallengeWitness` and have multiple inputs and outputs. I believe we can align the inputs and outputs so that at input[i] / output[i] you'll find the roots of the trees into which the inputs and outputs were inserted, with respect to **that input's sighash flag**. So if on the vPSBT level we have 10 vInputs and 10 vOutputs, if the first (index `0`) vInput wants to use a sighash that commits to a single input and all the outputs then on the dummy tx level we'd have `input[0]` be produced by inserting to an ms-smt just `vInput[0]` and `output[0]` would be produced by inserting all outputs in an ms-smt.

Can we go even further than that and do a "raw" mapping?

We can have each input in the dummy tx be produced by inserting into an ms-smt just the one vInput that has the same index. We can do exactly the same for the outputs, with each one being produced by an ms-smt that only contains the single vOutput on the same index. At that point can we just leverage the sighashes on the btc level? Nobody knows that the inputs/outputs are dummy, but the signatures can normally sign them in any sighash fashion already available to us via the backend.

If the latter works I believe it's the preferable approach?
LMK what you think

Reply:
I don't think this is an issue and is very likely just a misunderstanding around the `ChallengeWitness`.
That field is a special field that is only set in ownership proof files. An ownership proof is only created if you want/need to prove to someone that you do have the asset-level key to spend an asset. So it is just a signature showing you _could_ spend the asset, without actually spending it (so it's not a valid transfer proof).

So the field is only set for the output of `ProveAssetOwnership` and can only be verified with `VerifyAssetOwnership`.
The 1-in-1 out mentioned there is actually on the vPSBT level (not the dummy TX level), so it won't be affected by any changes we make due to the sighash implementation. Also, for the asset ownership proof we'll likely always use `SigHashDefault`.

Reply:
> We can have each input in the dummy tx be produced by inserting into an ms-smt just the one vInput that has the same index. We can do exactly the same for the outputs, with each one being produced by an ms-smt that only contains the single vOutput on the same index. At that point can we just leverage the sighashes on the btc level? Nobody knows that the inputs/outputs are dummy, but the signatures can normally apply to them in any sighash fashion already available to us via the backend.

Yes, I think that is what we want. We'd do a raw mapping and then just use the BTC-VM level sighash functionality.